x-pack/plugins/monitoring/init.js

```diff
@@ -4,42 +4,86 @@
  */
 export const init = (monitoringPlugin, server) => {
-  monitoringPlugin.status.yellow('Initializing');
-  const xpackMainPlugin = server.plugins.xpack_main;
+  const config = server.config();
 
-  xpackMainPlugin.status.once('green', async () => {
-    const config = server.config();
+  const collectorSet = new CollectorSet(server);
+  server.expose('collectorSet', collectorSet); // expose the collectorSet service
+  /*
+   * Register collector objects for stats to show up in the APIs
+   */
+  collectorSet.register(getOpsStatsCollector(server));
+  collectorSet.register(getKibanaUsageCollector(server));
+  collectorSet.register(getSettingsCollector(server));
+
+  /*
+   * Instantiate and start the internal background task that calls collector
+   * fetch methods and uploads to the ES monitoring bulk endpoint
+   */
+  const xpackMainPlugin = server.plugins.xpack_main;
+  xpackMainPlugin.status.once('green', async () => { // first time xpack_main turns green
+    /*
+     * End-user-facing services
+     */
     const uiEnabled = config.get('xpack.monitoring.ui.enabled');
 
     if (uiEnabled) {
       await instantiateClient(server); // Instantiate the dedicated ES client
       await initMonitoringXpackInfo(server); // Route handlers depend on this for xpackInfo
       await requireUIRoutes(server);
     }
+  });
 
-    if (config.get('xpack.monitoring.kibana.collection.enabled')) {
-      const collectorSet = createCollectorSet(monitoringPlugin.kbnServer, server); // instantiate an object for collecting/sending metrics and usage stats
-      server.expose('collectorSet', collectorSet); // expose the collector set object on the server. other plugins will call statsCollectors.register(collector) to define their own collection
-    }
+  const bulkUploader = initBulkUploader(monitoringPlugin.kbnServer, server);
+  const kibanaCollectionEnabled = config.get('xpack.monitoring.kibana.collection.enabled');
+  const { info: xpackMainInfo } = xpackMainPlugin;
+
+  if (kibanaCollectionEnabled) {
+    /*
+     * Bulk uploading of Kibana stats
+     */
+    xpackMainInfo.onLicenseInfoChange(() => {
+      // use updated xpack license info to start/stop bulk upload
+      const mainMonitoring = xpackMainInfo.feature('monitoring');
+      const monitoringBulkEnabled = mainMonitoring && mainMonitoring.isAvailable() && mainMonitoring.isEnabled();
+      if (monitoringBulkEnabled) {
+        bulkUploader.start(collectorSet);
+      } else {
+        bulkUploader.handleNotEnabled();
+      }
+    });
+  } else if (!kibanaCollectionEnabled) {
+    server.log(
+      ['info', LOGGING_TAG, KIBANA_MONITORING_LOGGING_TAG],
+      'Internal collection for Kibana monitoring will is disabled per configuration.'
+    );
+  }
 
-    monitoringPlugin.status.green('Ready');
+  // xpack main plugin status goes red if ES client lost connection
+  xpackMainPlugin.status.on('red', () => {
+    bulkUploader.handleConnectionLost();
```

Comment:
I think this is unnecessary and will cause problems because we are stopping the bulkUploader without starting again. Instead I think the bulk uploader should just try to talk to elasticsearch and fail if it can't. If we really want to get rid of the errors while Kibana is a red status we can check `xpackMainPlugin.status.state !== 'red'` before logging about errors talking to elasticsearch or something, but I suggest avoiding toggling the bulkUploader based on plugin status.

Reply:
> stopping the bulkUploader without starting again

It will get restarted when the the xpack main plugin status change causes the license info to change, but yeah - that is a pretty indirect signal.

> we can check `xpackMainPlugin.status.state !== 'red'`

I would have liked to not log avoidable errors, and checking the xpack main plugin state feels like something that should be handled by the platform with. Earlier in this code, I was passing `xpackMainInfo` into the bulk uploader constructor, and I really like that I'm not doing that in the latest version.

It sounds like checking the status to avoid logging errors is a nice-to-have. I like the idea of simplifying the code, so I'll just remove this status handler.